feat: use jujuparams and not jujucloud

ale8k opened the pull request:
## Description
In my original PR, we used jujucloud.CloudCredential (because it was easier), it enabled us to send a set of credentials and not worry about which the user has picked as juju will iterate over them. Now, we're explicitly sending a single credential and taking the first. 

It is picking the default credential unless otherwise specified by a new --credential flag.


This also needs Q/A'ing before merging.  So do not merge this until it is QA'd. Once #1682 is merged, this can be rebased with v3 and QA'd by updating the env to point at JIMM.

## Engineering checklist
<!-- *Check only items that apply* -->

- [ ] Documentation updated
- [ ] Covered by unit tests
- [ ] Covered by integration tests

## Test instructions
<!-- *(optional)* Describe any non-standard test instructions and configuration settings. Delete this section if not applicable. -->

ale8k committed:
feat: use jujuparams and not jujucloud

kian99 reviewed (commented):
I would just be explicit here, have the user pick the credential they want to use so command becomes something like `juju jaas bootstrap <other-position-args> --credential foo`. I would also make it a flag rather than an argument because too many arguments becomes very hard to read.

ale8k commented:
> I would just be explicit here, have the user pick the credential they want to use so command becomes something like `juju jaas bootstrap <other-position-args> --credential foo`. I would also make it a flag rather than an argument because too many arguments becomes very hard to read.

We can default to first and also add the flag - maybe that's best? I'll update this PR with the flag


I've updated it to just use default and not first (which I think is maybe right?) and also added a --credential flag.

ale8k commented on internal/jujuapi/jimm.go:
```diff
@@ -643,6 +644,14 @@ func (r *controllerRoot) BootstrapStart(ctx context.Context, req apiparams.Boots
+	)
+
+	cloudCred := cloud.NewEmptyCloudCredential()
+	cloudCred.AuthCredentials["default"] = in
```
This "might" be wrong, me and Joe did check it out about using the term default and not setting DefaultCredential explicitly and it seems ok, but QA should discover if this is wrong and the string explicitly needs setting.

kian99 replied:
I'd also like to see some more investigation before we land stuff that we only think might work. But considering it's still a wip, I'll aprove.

ale8k committed:
chore: allow user to specify credential

luci1900 reviewed (approved):
I like explicit.

SimoneDutto commented on internal/jujuapi/jimm.go:
```diff
@@ -643,6 +644,14 @@ func (r *controllerRoot) BootstrapStart(ctx context.Context, req apiparams.Boots
+	)
+
+	cloudCred := cloud.NewEmptyCloudCredential()
+	cloudCred.AuthCredentials["default"] = in
```
can we extract "default" to a const of some sort?

ale8k replied:
Juju doesn't use a const :(. You can also name it whatever you like as long as the default cred string is set. I can't see in the code what it does though if it isn't set, so I need to investigate that (otherwise I'm just gonna explicitly set it to default and create a const on our side.

SimoneDutto reviewed (commented):
when we say "it might need QA" why aren't we doing the qa before proposing the PR? Otherwise we need to merge this out of trust and then we need another pr to fix it

Maybe there is something blocking the QA, i dont' know

ale8k commented:
This is proving to be difficult, going to close it and come back later.

kian99 commented:
> This is proving to be difficult, going to close it and come back later.

Do we have a follow-up Jira card for this?

ale8k commented:
> > This is proving to be difficult, going to close it and come back later.
> 
> Do we have a follow-up Jira card for this?

I'm making one now.

ale8k commented:
Here: https://warthogs.atlassian.net/browse/JUJU-8531
